pkg/controller/ec2/vpcendpoint/setup.go

```diff
@@ -0,0 +1,384 @@
+package vpcendpoint
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/nsf/jsondiff"
+	"github.com/pkg/errors"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	svcsdk "github.com/aws/aws-sdk-go/service/ec2"
+	svcsdkapi "github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	cpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/ec2/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// SetupVPCEndpoint adds a controller that reconciles VPCEndpoint.
+func SetupVPCEndpoint(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(svcapitypes.VPCEndpointGroupKind)
+	opts := []option{setupExternal}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&svcapitypes.VPCEndpoint{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.VPCEndpointGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func setupExternal(e *external) {
+	c := &custom{client: e.client, kube: e.kube}
+	e.delete = c.delete
+	e.preCreate = preCreate
+	e.postCreate = postCreate
+	e.postObserve = postObserve
+	e.isUpToDate = isUpToDate
+	e.preUpdate = c.preUpdate
+	e.postUpdate = postUpdate
+	e.filterList = filterList
+}
+
+type custom struct {
+	kube   client.Client
+	client svcsdkapi.EC2API
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.VPCEndpoint, obj *svcsdk.CreateVpcEndpointInput) error {
+	// set external name as tag on the vpc endpoint
+	resType := "vpc-endpoint"
+	key := "Name"
+	value := meta.GetExternalName(cr)
+
+	// Clear SGs, RTs, and Subnets if they're empty
+	if cr.Spec.ForProvider.SecurityGroupIDs == nil || len(cr.Spec.ForProvider.SecurityGroupIDs) == 0 {
```

Comment:
I believe we could drop these nil checks: https://medium.com/@habibridho/golang-nil-vs-empty-slice-87fd51c0a4d

Reply:
Nil checks are dropped for this.